runtime/src/main/java/com/dylibso/chicory/runtime/TypeValidator.java

```diff
@@ -6,28 +6,45 @@
 import com.dylibso.chicory.wasm.types.Instruction;
 import com.dylibso.chicory.wasm.types.OpCode;
 import com.dylibso.chicory.wasm.types.ValueType;
-import java.util.ArrayDeque;
-import java.util.Deque;
+import java.util.ArrayList;
 import java.util.List;
 
 // Heavily inspired by wazero
 // https://github.com/tetratelabs/wazero/blob/5a8a053bff0ae795b264de9672016745cb842070/internal/wasm/func_validation.go
 public class TypeValidator {
 
-    private Deque<ValueType> valueTypeStack = new ArrayDeque<>();
-    private Deque<Integer> stackLimit = new ArrayDeque<>();
-    private Deque<List<ValueType>> returns = new ArrayDeque<>();
-    private Deque<Deque<ValueType>> unwindStack = new ArrayDeque<>();
+    private List<ValueType> valueTypeStack = new ArrayList<>();
+    private List<List<ValueType>> returns = new ArrayList<>();
+    private List<List<ValueType>> prevStack = new ArrayList<>();
+
+    private static <T> T peek(List<T> list) {
+        return list.get(list.size() - 1);
+    }
+
+    private static <T> T pop(List<T> list) {
+        var val = list.get(list.size() - 1);
+        list.remove(list.size() - 1);
+        return val;
+    }
+
+    private static <T> void push(List<T> list, T elem) {
+        list.add(elem);
+    }
+
+    private static <T> List<T> clone(List<T> list) {
+        var res = new ArrayList<T>();
+        res.addAll(list);
+        return res;
```

Comment:
```suggestion
        return new ArrayList<>(list);
```

Reply:
I'm not even sure if this deserves a separated method tbh.

Reply:
Possibly you are right @mariofusco , just a lot of forth and back to get there, I'll review the entire class at the end, there are other things to be tightened up ...